--- a/arch/arm/boot/dts/suniv-f1c100s.dtsi	2020-01-04 00:33:54.256666960 -0600
+++ b/arch/arm/boot/dts/suniv-f1c100s.dtsi	2020-01-03 17:00:15.331819585 -0600
@@ -6,6 +6,7 @@
 
 #include <dt-bindings/clock/suniv-ccu-f1c100s.h>
 #include <dt-bindings/reset/suniv-ccu-f1c100s.h>
+#include <dt-bindings/dma/sun4i-a10.h>
 
 / {
 	#address-cells = <1>;
@@ -71,6 +73,20 @@
 			};
 		};
 
+		codec: codec@1c23c00 {
+			#sound-dai-cells = <0>;
+			compatible = "allwinner,suniv-f1c100s-codec";
+			reg = <0x01c23c00 0x400>;
+			interrupts = <21>;
+			clocks = <&ccu CLK_BUS_CODEC>, <&ccu CLK_CODEC>;
+			clock-names = "apb", "codec";
+			dmas = <&dma SUN4I_DMA_NORMAL 12>,
+			       <&dma SUN4I_DMA_NORMAL 12>;
+			dma-names = "rx", "tx";
+			resets = <&ccu RST_BUS_CODEC>;
+			status = "disabled";
+		};
+
 		spi0: spi@1c05000 {
 			compatible = "allwinner,suniv-f1c100s-spi",
 				     "allwinner,sun8i-h3-spi";
@@ -175,6 +191,16 @@
 			#size-cells = <0>;
 		};
 
+		dma: dma-controller@1c02000 {
+			compatible = "allwinner,suniv-f1c100s-dma";
+			reg = <0x01c02000 0x1000>;
+			interrupts = <18>;
+			clocks = <&ccu CLK_BUS_DMA>;
+			resets = <&ccu RST_BUS_DMA>;
+			#dma-cells = <2>;
+		};
+
+
 		ccu: clock@1c20000 {
 			compatible = "allwinner,suniv-f1c100s-ccu";
 			reg = <0x01c20000 0x400>;
--- a/arch/arm/boot/dts/suniv-f1c100s-licheepi-nano.dts	2020-01-04 00:33:54.256666960 -0600
+++ b/arch/arm/boot/dts/suniv-f1c100s-licheepi-nano.dts	2020-01-01 21:42:10.770126979 -0600
@@ -69,3 +69,11 @@
 	usb0_id_det-gpio = <&pio 4 2 GPIO_ACTIVE_HIGH>; /* PE2 */
 	status = "okay";
 };
+
+&codec {
+	allwinner,audio-routing =
+		"Headphone", "HP",
+		"Headphone", "HPCOM",
+		"MIC", "Mic";
+	status = "okay";
+};
